### PR TITLE
Faster canonicalization

### PR DESCRIPTION
**Pull request by robert3005**

### CODE REVIEW

File: vortex-array/src/array/chunked/canonical.rs [hunk 1]
```diff
@@ -166,13 +166,13 @@ fn swizzle_struct_chunks(
 fn pack_bools(chunks: &[Array], nullability: Nullability) -> VortexResult<BoolArray> {
     let len = chunks.iter().map(|chunk| chunk.len()).sum();
     let validity = validity_from_chunks(chunks, nullability);
-    let mut bools = Vec::with_capacity(len);
+    let mut buffer = BooleanBufferBuilder::new(len);
```
Comment (robert3005) [inline vortex-array/src/array/chunked/canonical.rs:174]: building BooleanBuffer avoids copy compared to creating BooleanBuffer from Vec<bool>

File: vortex-array/src/array/chunked/canonical.rs [hunk 2]
```diff
@@ -210,19 +210,37 @@ fn pack_varbin(
...
-    let len = chunks.iter().map(|chunk| chunk.len()).sum();
-    let mut builder = VarBinBuilder::<i32>::with_capacity(len);
+    let len: usize = chunks.iter().map(|c| c.len()).sum();
+    let mut offsets = Vec::with_capacity(len + 1);
+    offsets.push(0);
+    let mut buffer = Vec::new();
 
     for chunk in chunks {
         let chunk = chunk.clone().into_varbin()?;
-        chunk.with_iterator(|iter| {
```
Comment (robert3005) [inline vortex-array/src/array/chunked/canonical.rs:218]: While this method is convenient it's not very fast if you know you just want to copy the values

File: vortex-array/src/array/chunked/canonical.rs [hunk 3]
```diff
@@ -205,33 +209,34 @@ fn pack_primitives(
...
+                .skip(1)
+                .map(|off| *off + offset_adjustment),
+        );
+        buffer.extend_from_slice(chunk.bytes().into_primitive()?.buffer());
     }
 
-    Ok(builder.finish(dtype.clone()))
-}
-
-fn validity_from_chunks(chunks: &[Array], nullability: Nullability) -> Validity {
```
Comment (robert3005) [inline vortex-array/src/array/chunked/canonical.rs:228]: this is just logicalvalidity of ChunkedArray

**Review by AdamGS — APPROVED** [review]

File: vortex-array/src/array/chunked/canonical.rs [hunk 4]
```diff
@@ -163,16 +169,15 @@ fn swizzle_struct_chunks(
 ///
 /// It is expected this function is only called from [try_canonicalize_chunks], and thus all chunks have
 /// been checked to have the same DType already.
-fn pack_bools(chunks: &[Array], nullability: Nullability) -> VortexResult<BoolArray> {
+fn pack_bools(chunks: &[Array], validity: Validity) -> VortexResult<BoolArray> {
     let len = chunks.iter().map(|chunk| chunk.len()).sum();
-    let validity = validity_from_chunks(chunks, nullability);
-    let mut bools = Vec::with_capacity(len);
+    let mut buffer = BooleanBufferBuilder::new(len);
```
Comment (AdamGS) [inline vortex-array/src/array/chunked/canonical.rs:174]: this change was on my list :)

**Pull request merged**